src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java

```diff
@@ -2,25 +2,26 @@
 
 import pl.cleankod.exchange.core.domain.Money;
 import pl.cleankod.exchange.core.gateway.CurrencyConversionService;
-import pl.cleankod.exchange.provider.nbp.ExchangeRatesNbpClient;
+import pl.cleankod.exchange.provider.nbp.ExchangeRatesNbpClientService;
 import pl.cleankod.exchange.provider.nbp.model.RateWrapper;
 
 import java.math.BigDecimal;
 import java.math.RoundingMode;
 import java.util.Currency;
 
 public class CurrencyConversionNbpService implements CurrencyConversionService {
-    private final ExchangeRatesNbpClient exchangeRatesNbpClient;
+    private final ExchangeRatesNbpClientService exchangeRatesNbpClientService;
 
-    public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClient) {
-        this.exchangeRatesNbpClient = exchangeRatesNbpClient;
+    public CurrencyConversionNbpService(ExchangeRatesNbpClientService exchangeRatesNbpClientService) {
+        this.exchangeRatesNbpClientService = exchangeRatesNbpClientService;
     }
 
     @Override
     public Money convert(Money money, Currency targetCurrency) {
-        RateWrapper rateWrapper = exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode());
+        RateWrapper rateWrapper = exchangeRatesNbpClientService.fetch("A", targetCurrency.getCurrencyCode());
         BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
+        //HALF_EVEN ala bankers rounding, not to lose money over time (hopefully not a code smell comment)
+        BigDecimal calculatedRate = money.amount().divide(midRate, 2, RoundingMode.HALF_EVEN);
```

Comment:
This is an interesting solution. Aren't there still situations where fractions are lost? What can we do to remove this responsibility from our app (should we?)?

Reply:
Fractions are definitely lost on rounding. HALF_EVEN is intended to avoid bias accumulation over time.  Ownership and responsibility wise it seems to be more of a business question rather than technical. Given we are a currency-rate-convertor it sounds we should have the ownership.